system_design/design_web_crawler.md

```diff
@@ -0,0 +1,228 @@
+
+# Web Crawler
+
+**Uses**
+
+Web crawlers are useful for accomplishing the following tasks.
+- How search engines get their data (ex. Google, Baidu)
+- Test web page links
+- Monitor content or structure updates on a page
+- Mirroring 
+    - mirroring: making a copy of a website
+- Copyright infrigement checks
+
+This specific solution does not attempt to address both the crawler AND a use case above. It is only attempting to address the crawler design.
+
+**Design Specifications Needed to Pass this Interview**
+Here are the specific sections you will need to address in a web crawler system design interview.
+1. [Ask questions](#ask-questions)
+2. [Functional Requirements](#functional-requirements)
+3. [Non-functional Requirements](#non-functional-requirements)
+4. [Estimation](#estimation)
+5. [Diagram the System](#diagram-the-system)
+6. [Define Data Models](#define-data-models)
+7. [Define APIs](#define-apis)
+8. [Deep Dive on Diagram Section](#deep-dive-on-diagram-section)
+9. [Faults and Failures](#faults-and-failures)
+
+## Ask questions
+
+You are asking questions to clarify the requirements and constraings of the system you are building.
+
+- **Goal**:
+    - Do I want to index all web pages?
+    - Do I want to index all pages within a specific domain?
+    - Do I want to index specific pages within a specific domain?
+    - Do I want to index specific pages across domains?
+- **Scale**
+    - How many web pages will be crawled?
+    - How often should these pages be crawled?
+    - (Do different pages have different crawl rates?)
```

Comment:
Why have just this line in brackets?

Reply:
I believe I'm signaling to myself to only ask if time permits OR if I need to update estimations later. 

- [ ] remove brackets